draft-ietf-quic-transport.md

```diff
@@ -2978,6 +2897,89 @@ The details of loss detection and congestion control are described in
 {{QUIC-RECOVERY}}.
 
 
+## Explicit Congestion Notification {#ecn}
+
+QUIC endpoints use Explicit Congestion Notification (ECN) {{!RFC3168}} to detect
+and respond to network congestion.  ECN allows a network node to indicate
+congestion in the network by setting a codepoint in the IP header of a packet
+instead of dropping it.  Endpoints react to congestion by reducing their sending
+rate in response, as described in {{QUIC-RECOVERY}}.
+
+To use ECN, QUIC endpoints first determine whether a path supports ECN marking
+and the peer is able to access the ECN codepoint in the IP header.  A network
+path does not support ECN if ECN marked packets get dropped or ECN markings are
+rewritten on the path. An endpoint verifies the path, both during connection
+establishment and when migrating to a new path (see {{migration}}).
+
+
+### ECN Counters
+
+On receiving a packet with an ECT or CE codepoint, an endpoint that can access
+the IP ECN codepoints increases the corresponding ECT(0), ECT(1), or CE count,
+and includes these counters in subsequent (see {{processing-and-ack}}) ACK
+frames (see {{frame-ack}}).
+
+A packet detected by a receiver as a duplicate does not affect the receiver's
+local ECN codepoint counts; see ({{security-ecn}}) for relevant security
+concerns.
+
+If an endpoint receives a packet without an ECT or CE codepoint, it responds per
+{{processing-and-ack}} with an ACK frame.  If an endpoint does not have access
+to received ECN codepoints, it acknowledges received packets per
+{{processing-and-ack}} with an ACK frame.
+
+
+### ECN Verification
+
+Each endpoint independently verifies and enables use of ECN by setting the IP
```

Comment:
I think we should say something about when this verification can be done. It can be done any time during the connection, but it might make the most sense to do it along with the handshake. We may want to add some retransmission suggestions: if a handshake packet is detected as lost then send the retransmission without ECT bits. (Feel free to add a TODO here if you don't want to address this in this PR.)

I'm also wondering if we should add a fwd reference in the handshake section to here, with something like "An endpoint MAY combine the cryptographic handshake with ECN verification, as described in {{}}".

Reply:
Happy to take other PRs as a follow-on.  Note that this is literally reordering text, so correcting any omission is strictly improvements on this.